### PR TITLE
Support for rai livestreams

### PR DESCRIPTION
**Pull request by rabblac**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information
This Pull Requests aims to support rai live streams listed [here](http://www.raiplay.it/dirette/).

On youtube-dl seems to work fine, but other players try to get index m3u8 the next fragments, which leads to an access denied error. (if you need them, [mpv logs with the two streams](https://gist.github.com/rabblac/0f6506da4ec37a9ae119d9dee5d5770b))

### CODE REVIEW

**Comment by dstftw** [comment]
Add as separate extractor.

**Comment by gfabiano** [comment]
yeah @rabblac , I was also working on this extractor and I noticed that mpv does not handle it good. I don't understand why because it uses ffmpeg as youtube-dl. The only solution for now i found is to use pipes:
`youtube-dl http://raiplay.it/dirette/rai3 -o - | mpv -`

**Comment by rabblac** [comment]
That's why mpv asks every time for the index m3u8 for new fragments, meanwhile youtube-dl increments by one the fragment.
Anyway, thanks for the fast reply dstftw, I'm working on this.

File: youtube_dl/extractor/rai.py [hunk 1]
```diff
@@ -218,7 +208,60 @@ def _real_extract(self, url):
...
+            'description': media.get('description'),
+            'uploader': media.get('channel'),
+            'creator': media.get('editor'),
+            'timestamp': timestamp,
+            'thumbnails': thumbnails,
+            'subtitles': subtitles,
+        }
 
+        info.update(relinker_info)
         return info
```
Comment (gfabiano) [inline youtube_dl/extractor/rai.py]: It's not better in this way?
```python
class RaiPlayLiveIE(RaiBaseIE):
    _VALID_URL = r'(?P<url>https?://(?:www\.)?raiplay\.it/dirette/(?P<id>\w*))'
    _TEST = {
        'url': 'http://www.raiplay.it/dirette/rai3',
        'only_matching': True,
    }

    def _real_extract(self, url):
        mobj = re.match(self._VALID_URL, url)
        url, channel = mobj.group('url', 'id')

        webpage = self._download_webpage(url, channel)
        re_id = r'<div([^>]*)data-uniquename=(["\'])[\w-]*(?P<id>%s)(\2)([^>]*?)>' % RaiBaseIE._UUID_RE
        video_id = self._html_search_regex(re_id, webpage, 'livestream-id', group='id')

        return {
            '_type': 'url_transparent',
            'url': 'http://www.raiplay.it/dirette/ContentItem-%s.html' % video_id,
            'ie_key': RaiPlayIE.ie_key()
        }
```

Comment (rabblac) [inline youtube_dl/extractor/rai.py]: It works well, so why not? Let's see what @dstftw says.

File: youtube_dl/extractor/rai.py [hunk 2]
```diff
@@ -208,7 +208,60 @@ def _real_extract(self, url):
...
+            'alt_title': media.get('subtitle'),
+            'description': media.get('description'),
+            'uploader': media.get('channel'),
+            'creator': media.get('editor'),
+            'timestamp': timestamp,
+            'thumbnails': thumbnails,
+            'subtitles': subtitles,
+        }
+
+        info.update(relinker_info)
```
Comment (dstftw) [inline youtube_dl/extractor/rai.py]: 230-264 no copy pastes.

**Comment by dstftw** [comment]
As already mentioned by @gfabiano you should **delegate** extraction to `RaiPlayIE` **not copy paste code**.

**Comment by rabblac** [comment]
Sure thing, didn't know that was possible. Commit sent.

File: youtube_dl/extractor/rai.py [hunk 3]
```diff
@@ -208,10 +208,31 @@ def _real_extract(self, url):
...
+class RaiPlayLiveIE(RaiBaseIE):
+    _VALID_URL = r'(?P<url>https?://(?:www\.)?raiplay\.it/dirette/(?P<id>\w*))'
+    _TEST = {
+        'url': 'http://www.raiplay.it/dirette/rai3',
+        'only_matching': True,
+    }
+
+    def _real_extract(self, url):
+        mobj = re.match(self._VALID_URL, url)
+        url, channel = mobj.group('url', 'id')
```
Comment (dstftw) [inline youtube_dl/extractor/rai.py]: What's the point capturing `url` when you already have it in `url`. Rewrite with `_match_id`.

File: youtube_dl/extractor/rai.py [hunk 4]
```diff
@@ -208,10 +208,31 @@ def _real_extract(self, url):
...
+
+        webpage = self._download_webpage(url, channel)
+        re_id = r'<div([^>]*)data-uniquename=(["\'])[\w-]*(?P<id>%s)(\2)([^>]*?)>' % RaiBaseIE._UUID_RE
+        video_id = self._html_search_regex(re_id, webpage, 'livestream-id', group='id')
+
+        return {
+            '_type': 'url_transparent',
+            'url': 'http://www.raiplay.it/dirette/ContentItem-%s.html' % video_id,
+            'ie_key': RaiPlayIE.ie_key()
```
Comment (dstftw) [inline youtube_dl/extractor/rai.py]: Use `url_result` if you don't adjust any metadata.

File: youtube_dl/extractor/rai.py [hunk 5]
```diff
@@ -156,7 +156,7 @@ class RaiPlayIE(RaiBaseIE):
         },
         'params': {
             'skip_download': True,
-        },
+        }
```
Comment (dstftw) [inline youtube_dl/extractor/rai.py]: Remove all pointless changes.

File: youtube_dl/extractor/rai.py [hunk 6]
```diff
@@ -208,10 +208,27 @@ def _real_extract(self, url):
         }
 
         info.update(relinker_info)
-
         return info
 
 
+class RaiPlayLiveIE(RaiBaseIE):
+    _VALID_URL = r'https?://(?:www\.)?raiplay\.it/dirette/(?P<id>\w*)'
```
Comment (dstftw) [inline youtube_dl/extractor/rai.py]: http://www.raiplay.it/dirette/ should not be matched.

Comment (gfabiano) [inline youtube_dl/extractor/rai.py]: You need only to change `(?P<id>\w*)` with `(?P<id>\w+)`

Comment (rabblac) [inline youtube_dl/extractor/rai.py]: Thanks

**Pull request closed without merging**